GoogleCloudPlatform/bank-of-anthos
# Generate kustomize/base in Release Automation

* At the moment, this pull-request targets the `kustomize-cloudsql` branch.
* Reason: to see the delta that this pull-request introduces.
* This pull-request will target `main` after we merge https://github.com/GoogleCloudPlatform/bank-of-anthos/pull/743.

### Background 

* `/release/make-release.sh` is run every time a new Bank of Anthos release is created.
* `/release/make-release.sh` currently generates the files in [/kubernetes-manifests](https://github.com/GoogleCloudPlatform/bank-of-anthos/tree/release/v0.5.4/kubernetes-manifests) by copying the files in [/dev-kubernetes-manifests](https://github.com/GoogleCloudPlatform/bank-of-anthos/tree/release/v0.5.4/dev-kubernetes-manifests).

### Change Summary

* Edit `/release/make-release.sh` such that it generates and git-commits the `/kustomize/base` files.
* Add a `kustomization.yaml` file to `/dev-kubernetes-manifests/`. This does **not** break `skaffold dev`. I've manually tested. 👍 

### How to Review

0. Use **Cloud Shell**! Your local machine will give you trouble (i.e., with `realpath` and `find ... sed ...` commands).

1. Make sure I've listed all the YAML files in `/dev-kubernetes-manifests/` inside `/dev-kubernetes-manifests/kustomization.yaml`.

2. Checkout this branch.
```bash
git clone https://github.com/GoogleCloudPlatform/bank-of-anthos.git
cd bank-of-anthos
git checkout kustomize-release-automation
``` 

3. Make sure the `/dev-kubernetes-manifests/kustomization.yaml` file builds.
```bash
kustomize build dev-kubernetes-manifests
```

You will see a concatenation of all the YAML files in `/dev-kubernetes-manifests`.

4. Comment out the last 7 commands from `make-release.sh`:

```bash
# git commit -m "release/${NEW_VERSION}"

# add tag
# git tag "${NEW_VERSION}"

# build and push release images
# skaffold config set local-cluster false
# skaffold build --default-repo="${REPO_PREFIX}" --tag="${NEW_VERSION}"
# skaffold config unset local-cluster

# push to repo
# git push --set-upstream origin "release/${NEW_VERSION}"
# git push --tags
```

5. Comment out the "no uncommited changes" requirement from `make-release.sh`:
```bash
# ensure there are no uncommitted changes
# if [[ $(git status -s | wc -l) -gt 0 ]]; then
#    echo "error: can't have uncommitted changes"
#    exit 1
# fi
```

6. Set environment variables for testing `make-release.sh`.

```bash
export NEW_VERSION=v9.9.9
```

7. Run `./make-release.sh`.

```bash
cd release
./make-release.sh
```

If you don't have the `realpath` command, just manually set `SCRIPT_DIR` inside `make-release.sh`.
Example:
```
SCRIPT_DIR="/home/nimjay/bank-of-anthos/release"
```

9. Make sure the `kustomize/base` files have been updated.
```bash
git diff --staged ../kustomize/base/*.yaml
```

You'll notice the replacement of `Copyright 2022` with `Copyright 2019`. I think this should be fine but let me know if you'd like me use `Copyright 2022` instead.

10. Make sure the `kustomize/base` updates have been staged via git.
```bash
git status
```

11. Clean up.
```bash
git reset --hard ../ # Revert all file changes (both staged YAMLs and make-release.sh).
git checkout main
git branch -D release/v9.9.9
```
